content/docs/testing-recipes.md

```diff
@@ -0,0 +1,403 @@
+  useEffect(
+    props => {
+      const timeoutID = setTimeout(() => {
+        props.onSelect(null);
+      }, 5000);
+      return () => {
+        clearTimeout(timeoutID);
+      };
+    },
+    [props.onSelect]
+  );
+  return [1, 2, 3, 4].map(choice => {
+    <button
+      key={choice}
+      data-test-id={choice}
+      onClick={() => props.onSelect(choice)}
+    >
+      {choice}
+    </button>;
+  });
+}
+```
+
+We can write tests for this component by leveraging jest's timer mocks, and testing the different states it can be in.
+
+```jsx
+// card.test.js
+jest.useFakeTimers();
+
+let container = null
+beforeEach(() => {
+  // setup a DOM element as a render target
+  container = document.createElement('div')
+  document.body.appendChild(container)
+})
+
+afterEach(() => {
+  // cleanup on exiting
+  container.remove();
+  container = null  
+})
+
+function render(ui){
+  act(() => {
+    ReactDOM.render(ui, container);
+  });
+}
+
+it("should select null after timing out", () => {
+  const yields = [];
+  render(<Card onSelect={selection => yields.push(selection)} />);
+  act(() => {
+    jest.advanceTimersByTime(100);
+  });
+  expect(yields).toBe([]);
+  act(() => {
+    jest.advanceTimersByTime(1000);
+  });
+  expect(yields).toBe([null]);
+});
+
+it("should cleanup on being removed", () => {
+  const yields = [];
+  render(
+    <Card onSelect={selection => yields.push(selection)} />
+  );
+  act(() => {
+    jest.advanceTimersByTime(100);
+  });
+  expect(yields).toBe([]);
+  act(() => {
+    render(null);
+  });
+  act(() => {
+    jest.advanceTimersByTime(1000);
+  });
+  expect(yields).toBe([]);
+});
+
+it("should accept selections", () => {
+  const yields = [];
+  render(
+    <Card onSelect={selection => yields.push(selection)} />
+  );
+
+  act(() => {
+    container
+      .querySelector("[data-test-id=2]")
+      .dispatchEvent(new MouseEvent("click", { bubbles: true }));
+  });
+
+  expect(yields).toBe([2]);
+});
+```
+
+### Snapshot testing {#snapshot-testing}
+
+Frameworks like Jest also let you save 'snapshots' of data. Combined with react-test-renderer, we can save component structures, and verify they don't break in the future. It's typically better to make more specific assertions than to use snapshots (or at least ensure that the snapshot is small) because these kinds of tests include implementation details meaning they break easily and teams can get desensitized to snapshot breakages.
+
+```jsx
+// hello.test.js, again
+import Hello from "./hello";
+import { create, act } from "react-test-renderer";
+
+// a helper to convert a React component tree to json
+function toJSON(element) {
+  let root;
+  act(() => {
+    root = create(element);
+  });
+  return root.toJSON();
+}
+
+it("should preserve it's structure", () => {
+  expect(toJSON(<Hello />)).toMatchSnapshot();
+  expect(toJSON(<Hello name="Sophie" />)).toMatchSnapshot();
+  expect(toJSON(<Hello name="Flarnie" />)).toMatchSnapshot();
+});
+```
+
```

Comment:
Maybe call out mocking a dependency as a common solution to avoid snapshots growing too much.

Reply:
I don't think dependencies should be mocked to prevent big snapshots. It could subtly break the code (eg - an effect in a child that called a callback would not be called any more, changing the observed result)

Reply:
ie - dependencies should only be mocked if there's a clear reason they don't belong in the test: they render a canvas, or can't run in test environments, etc. This is covered by the mocking-modules section.